crypto/o_str.c

```diff
@@ -90,6 +90,69 @@ size_t OPENSSL_strlcat(char *dst, const char *src, size_t size)
     return l + OPENSSL_strlcpy(dst, src, size);
 }
 
+/**
+ * @brief Converts a string to an unsigned long integer.
+ *
+ * This function attempts to convert a string representation of a number
+ * to an unsigned long integer, given a specified base. It also provides
+ * error checking and reports whether the conversion was successful.
+ * This function is just a wrapper around the POSIX strtoul function with
+ * additional error checking.  This implies that errno for the caller is set
+ * on calls to this function.
+ *
+ * @param str The string containing the representation of the number.
+ * @param endptr A pointer to a pointer to character. If not NULL, it is set
+ *               to the character immediately following the number in the
+ *               string.
+ * @param base The base to use for the conversion, which must be between 2,
+ *             and 36 inclusive, or be the special value 0. If the base is 0,
+ *             the actual base is determined by the format of the initial
+ *             characters of the string.
+ * @param num A pointer to an unsigned long where the result of the
+ *            conversion is stored.
+ *
+ * @return 1 if the conversion was successful, 0 otherwise. Conversion is
+ *         considered unsuccessful if no digits were consumed or if an error
+ *         occurred during conversion.
+ *
+ * @note It is the caller's responsibility to check if the conversion is
+ *       correct based on the expected consumption of the string as reported
+ *       by endptr.
+ */
+int OPENSSL_strtoul(const char *str, char **endptr, int base,
```

Comment:
```suggestion
int OPENSSL_strtoul(const char *str, unsigned long& num)
```
In my opinion, this wrapper function should have only two arguments as I suggested before. Otherwise, we might be overdoing things.

Reply:
I disagree.  what if a user explicitly wants to validate that an input value is in hex form rather than decimal form?  This is documented as a wrapper around strtoul, we should accept the same input parameters.

Reply:
My perspective was:

Firstly, that `strtoul` automatically detects the base for cases like hexadecimal. But I'm okay taking it as input argument.

Secondly, IMO, the `endptr` is certainly not required here. That we can manage internally as part of this function. considering our use case.

Reply:
> `strtoul` automatically detects the base for cases like hexadecimal

Not quite: if you give `base` as `10`, `strtoul` will fail with a `0x` prefix or any hex chars in the string. If you give `base` as `16`, leading `0x` is optional. If you give `base` as `0`, leading `0x` means hex, leading `0` means octal, otherwise base `10` is assumed